### ISSUE TITLE
Error in if (ncol(res) == 1) { : argument is of length zero

### ISSUE THREAD

**Issue by marboe123** [bgreenwell/fastshap#51]
Hello,

I receive the error above after I use this command:

`shap_values <- fastshap::explain(model_n, X = trainval, exact = TRUE)`

model_n is a multiclassification model fitted with xgboost.
trainval is my train-data

When I run this set-up for binary classification, the shap values are calculated correctly.
When I run this set-up for multiclass classification, the error above is generated.

Do you have any idea what can be the cause?

Thanks a lot!

**Comment by marboe123** [comment]
ps:

If I use the package SHAPforxgboost to calculate the shap values, I also receive an error:

```
Error in `colnames<-`(`*tmp*`, value = c(colnames(X_train), "BIAS")) : 
  attempt to set 'colnames' on an object with less than two dimensions

```

There are several other users who have the same error as well as can be seen at the bottem of this post:

https://liuyanguu.github.io/post/2019/07/18/visualization-of-shap-for-xgboost/

And here:

https://github.com/liuyanguu/SHAPforxgboost/issues/35

Maybe this could be an indication of the rootcause of the error.

Thank you.

**Comment by marboe123** [comment]
@bgreenwell thank you for your response! The output is :
![image](https://user-images.githubusercontent.com/32816338/196980520-7437167d-cb87-44f3-b2d0-25ae81df062a.png)

If I look into the lists I see BIAS, BIAS.1, BIAS.2 and similar for my variables: x, x.1, x.2 etc.

Are these the shap values I can use if the shap values at itself are sufficient for me?

**Comment by bgreenwell** [comment]
Yes! I’ll fix the package to account for multi class models. But here you have a list with one component of Shapley values for each of your three class outcomes. In the binary case you really only need one. I’ll leave this issue open until I can get a fix, but those are exactly what you’re looking for!

**Comment by marboe123** [comment]
That is great. I have already been testing with `predict(mymodel, newdata = data, predcontrib = TRUE)`.
Initially I did xgboost hyperparameter tuning on binary classification with fastshap-shapvalues with a 6GB GPU. This resulted in out of memory errors sometimes.
Next I moved to a 24GB GPU and I had no memory errors anymore.
Currently I changed to multiclass with shapvalues based on `predict(mymodel, newdata = data, predcontrib = TRUE)` and I do experience the memory limit error again.
Do you think it is possible that fastshap does need less memory than the predict method or is this not possible?
I will test my multiclass script without shapvalues as well to see if it will run without memory errors.
Thanks!

**Comment by bgreenwell** [comment]
For multiclass `xgboost`, `predict(..., predcontrib = TRUE)` returns a 3-D array `[n, p+1, num_class]` rather than a 2-D matrix. The old code tried to coerce this to a flat matrix, producing garbage output and then failing with a cryptic error. Similarly for multiclass `lightgbm`, which returns `(p+1) * num_class` columns.

The fix (in the latest commit) detects the multiclass case immediately and stops with an informative message:

> Exact Shapley values (`exact = TRUE`) are not yet supported for multiclass `xgboost` models. Use the Monte Carlo approximation (default) with a prediction wrapper that returns the class score(s) of interest instead.

The MC path (`exact = FALSE`, the default) works fine for multiclass — just write a `pred_wrapper` that returns the score for the class you care about (or all classes, if you call `explain()` once per class). Full multiclass support for `exact = TRUE` may be added in a future release.
